plugins/via-usbhub/fu-via-usbhub-device.c

```diff
@@ -0,0 +1,1275 @@
+	if (fu_device_has_custom_flag (device, "unlock_813") &&
+	    !fu_via_usbhub_device_spi_read_data (self, 0x0, buf_tmp, sizeof(buf_tmp), &error_tmp)) {
+		g_warning ("failed to read, trying to unlock 813: %s", error_tmp->message);
+		if (!fu_via_usbhub_device_vdr_unlock_813 (self, error))
+			return FALSE;
+		if (!fu_via_usbhub_device_spi_read_data (self, 0x0, buf_tmp, sizeof(buf_tmp), error)) {
+			g_prefix_error (error, "813 unlock fail: ");
+			return FALSE;
+		}
+		g_debug ("813 unlock OK");
+		/* VL813 & VL210 have same PID (0x0813), and only VL813 can reply */
+		self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL813;
+	} else {
+		if (!fu_via_usbhub_device_guess_kind (self, error))
+			return FALSE;
+	}
+
+	/* get the flash chip attached */
+	if (!fu_via_usbhub_device_spi_read_flash_id (self, &flash_id, error)) {
+		g_prefix_error (error, "failed to read SPI chip ID: ");
+		return FALSE;
+	}
+	if (flash_id != 0x0) {
+		g_autofree gchar *instance_id = NULL;
+		instance_id = g_strdup_printf ("SPI\\FLASH_%04X", flash_id);
+		g_debug ("using flash part %s", instance_id);
+		fu_device_add_instance_id (FU_DEVICE (self), instance_id);
+	}
+
+	/* detect update protocol from the first block of flash */
+	fw = fu_via_usbhub_device_dump_firmware (self, 0x20, error);
+	if (fw == NULL)
+		return FALSE;
+	buf = g_bytes_get_data (fw, NULL);
+	switch (buf[0]) {
+	/* VL810~VL813 */
+	case 0x0d:
+		self->update_protocol = 0x1;
+		self->disable_powersave = TRUE;
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+		break;
+	/* VL817~ */
+	case 0x05:
+		self->update_protocol = 0x2;
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_SELF_RECOVERY);
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+		break;
+	default:
+		g_warning ("unknown update protocol, buf[0]=0x%x", buf[0]);
+		break;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_via_usbhub_device_prepare_firmware (FuDevice *device,
+				       GBytes *fw,
+				       FwupdInstallFlags flags,
+				       GError **error)
+{
+	FuViaUsbhubDevice *self = FU_VIA_USBHUB_DEVICE (device);
+	FuViaUsbhubDeviceKind device_kind;
+	g_autoptr(FuFirmware) firmware = fu_via_usbhub_firmware_new ();
+
+	/* check size */
+	if (g_bytes_get_size (fw) < fu_device_get_firmware_size_min (device)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "firmware too small, got 0x%x, expected >= 0x%x",
+			     (guint) g_bytes_get_size (fw),
+			     (guint) fu_device_get_firmware_size_min (device));
+		return NULL;
+	}
+	if (g_bytes_get_size (fw) > fu_device_get_firmware_size_max (device)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "firmware too large, got 0x%x, expected <= 0x%x",
+			     (guint) g_bytes_get_size (fw),
+			     (guint) fu_device_get_firmware_size_max (device));
+		return NULL;
```

Comment:
aren't these checks handled by fu-device already?

Reply:
I don't think so; we'd need to do them on the GBytes that come out of the `FuFirmwareImage` rather than the `FuFirmware` to be valid in all cases. The firmware size stuff is all very confusing, so I'm waiting back from some more details from VIA.